README.md
# BNBT-AGMS
This is a fixed up version of the BNBT BitTorrent Tracker to get it to run on current day computers, which have larger numbers (32 or 64 bits) and newer code libraries.

A Tracker is a dual purpose server-side program that lists available torrents for download on a web page, and lets client programs find other client programs for file sharing.

I've started with the code from BNBT 8.5 as of July 2006, as found on https://sourceforge.net/projects/bnbtusermods/files/BNBT85/July%202006%20Critical%20Update/ in 20060727-bnbt85-src.zip  The documentation and sample files (without the precompiled executables) are from bnbt-85b-untested.zip, a slightly earlier version.

I'm trying to get the basic BNBT to compile under Fedora 28 Linux, a circa 2018 version of the GNU/Linux operating system.  There's an option for a MySQL database version of BNBT, which I'm ignoring, so that likely won't build.

\- AGMS20180930


# BNBT Beta 8.1 (a C++ BitTorrent tracker)

http://bnbt.go-dedicated.com/

BNBT was written by Trevor Hogan. BNBT is a complete port of the original Python BitTorrent tracker to C++ using the STL for data storage and basic network sockets for network communication. BNBT is fast, efficient, customizable, easy to use, powerful, and portable. BNBT is covered under the GNU Lesser General Public License (LGPL).

## DOCUMENTATION

THIS IS IMPORTANT! PLEASE READ THE DOCUMENTATION!

The documentation for this release is available in the bnbt/doc directory. Updated documentation is available at http://bnbt.go-dedicated.com/doc/.

## CHANGELOG

Beta 8.1
 - major optimizations (no more threads)
  * tracker links were removed since they were threaded
 - added 'fast cache' to speed up index/scrape generation at a cost of delayed stats
  * added the "bnbt_refresh_fast_cache_interval" config value
  * the tracker will cache index/scrape data for at most this many seconds
 - slightly better error messages (thanks Olaf)
 - fixed a bug where BNBT would report bad headers when none were sent
 - minor optimizations (e.g. global recv buffer)
 - minor MySQL query changes
 - modified MySQL tables to use BLOB instead of BINARY CHAR
 - removed XML dumping
 - added RSS dumping (thanks labarks)
  * added a ton of config values for RSS dumping
 - changed /torrent.html linking to /torrents/<hash>.torrent
 - no longer send compressed responses if size_of_compressed > size_of_raw
 - fixed some compiler warnings and removed -w flag from Makefile

Beta 8.0 Release 2
 - fixed a serious crash bug with tracker links on secondary trackers
 - fixed a bug where BNBT would constantly refresh the static header and footer (credit Elandal)
 - fixed a query error with stats dumping when mysql_override_dstate was off
 - mysql.h is now include as mysql/mysql.h on some systems
 - and stdio.h is now included by default on all systems
 - fixed a major bug where some internal databases would be needlessly copied and deleted
 - added the "bnbt_max_recv_size" config value

Beta 8.0
 - supressed connection reset errors
 - supressed deflate errors (on Z_OK)
 - added MySQL support
  * added the "mysql_host" config value
  * added the "mysql_database" config value
  * added the "mysql_user" config value
  * added the "mysql_password" config value
  * added the "mysql_port" config value
  * added the "mysql_refresh_allowed_interval" config value
  * added the "mysql_refresh_stats_interval" config value
  * added the "mysql_override_dstate" config value
  * added BNBTPHP (example PHP frontend)
 - added the "bnbt_robots_txt" config value
 - removed the zlib source code folder
 - fixed a bug where deleted torrents would sometimes recover their tag data on restart
 - externally linked torrents are now linked even more correctly
 - pruned the changelog


doc/changelog.txt
Beta 8.1 Release 3
 - fixed a major bug with compact announces

Beta 8.1
 - major optimizations (no more threads)
  * tracker links were removed since they were threaded
 - added 'fast cache' to speed up index/scrape generation at a cost of delayed stats
  * added the "bnbt_refresh_fast_cache_interval" config value
  * the tracker will cache index/scrape data for at most this many seconds
 - slightly better error messages (thanks Olaf)
 - fixed a bug where BNBT would report bad headers when none were sent
 - minor optimizations (e.g. global recv buffer)
 - minor MySQL query changes
 - modified MySQL tables to use BLOB instead of BINARY CHAR
 - removed XML dumping
 - added RSS dumping (thanks labarks)
  * added a ton of config values for RSS dumping
 - changed /torrent.html linking to /torrents/<hash>.torrent
 - no longer send compressed responses if size_of_compressed > size_of_raw
 - fixed some compiler warnings and removed -w flag from Makefile

Beta 8.0 Release 2
 - fixed a serious crash bug with tracker links on secondary trackers
 - fixed a bug where BNBT would constantly refresh the static header and footer (credit Elandal)
 - fixed a query error with stats dumping when mysql_override_dstate was off
 - mysql.h is now include as mysql/mysql.h on some systems
 - and stdio.h is now included by default on all systems
 - fixed a major bug where some internal databases would be needlessly copied and deleted
 - added the "bnbt_max_recv_size" config value

Beta 8.0
 - supressed connection reset errors
 - supressed deflate errors (on Z_OK)
 - added MySQL support
  * added the "mysql_host" config value
  * added the "mysql_database" config value
  * added the "mysql_user" config value
  * added the "mysql_password" config value
  * added the "mysql_port" config value
  * added the "mysql_refresh_allowed_interval" config value
  * added the "mysql_refresh_stats_interval" config value
  * added the "mysql_override_dstate" config value
  * added BNBTPHP (example PHP frontend)
 - added the "bnbt_robots_txt" config value
 - fixed a bug where deleted torrents would sometimes recover their tag data on restart
 - externally linked torrents are now linked even more correctly
 - various project file changes (e.g. dynamic linking on Windows)
 - pruned the changelog

Beta 7.7
 - added the "bnbt_name_length" config value
 - added the "bnbt_comment_length" config value
 - added tracker links (from BNBT Beta 7.3)
 - major optimizations
  * converted vector loops to use iterators
  * converted CAtomDicti to use STL maps
 - info hashes as well as names now link to the stats page

Beta 7.6
 - upgraded base64 decoding library
 - upgraded zlib to version 1.2.1
  * zlib now handles gzip compression headers and footers
  * changed the "bnbt_compression_level" config value to default to 6 instead of 5
 - started converting some isXXX virtual functions to dynamic casts
 - started converting some structures to STL maps
 - added request_t client request structure
 - added bCompressOK flag to response_t server response structure
 - moved some request processing to the client thread
 - some non fatal errors are now reported as warnings instead
 - fixed a major bug with FreeBSD
 - fixed a bug with the archive dir (credit MrMister)
 - added support for the compact announce extension
 - added the "bnbt_show_average_ul_rate" config value
 - added the "bnbt_show_average_dl_rate" config value
 - added support for displaying average peer rates
 - misc HTML changes

Beta 7.5
 - removed the command console
 - info link now defaults to a blank value instead of "http://"
 - fixed a bug where moderators couldn't see the 'delete torrent' link
 - added support for the min_request_interval scrape flag
 - added support for the no_peer_id announce extension
 - removed logging of "client timed out" messages
 - fixed a MAJOR algorithm error with /announce (sorry Bram)
 - fixed a bug when pthread_create fails

Beta 7.4
 - fixed a bug where usernames with non standard characters couldn't be edited or deleted
 - fixed a bug where config entries without a value (i.e. "port =") would crash on startup
 - ip addresses are now only displayed with comments posted by a guest
 - added the "bnbt_max_peers_display" config value
 - added the "bnbt_realm" config value
 - moderators can now delete torrents
 - added the 'downloaded' key to /scrape
 - the 'Your Torrents' list now displays torrent names if possible
 - added a 'DOWNLOAD TORRENT' link to the stats page
 - removed logging to console (messages are logged to file only)
 - added a basic command console
 - extensive optimizations
  * removed stealers due to conflicts with optimized code
  * removed tracker links due to conflicts with optimized code

Beta 7.3
 - announce errors are now returned as 200 OK with a 'failure reason' key
 - fixed a bug where some filter links weren't escaped correctly
 - added the "bnbt_force_announce_on_download" config value
 - added "name=" parameter to Content-Type on torrent download (should fix some torrent download problems)
 - added a "Count Unique Peers" link to the admin panel
 - added the "bnbt_tlink_server" config value
 - added the "bnbt_tlink_connect" config value
 - added the "bnbt_tlink_bind" config value
 - added the "bnbt_tlink_password" config value
 - added the "bnbt_tlink_port" config value
 - added tracker links (read link.txt)
 - changed the behaviour of the server to leave excess connections in the socket queue instead of closing
 - added the "bnbt_allow_info_link" config value
 - added an Info Link column
 - added the "bnbt_archive_dir" config value
 - compiling BNBT under Windows should be significantly faster (WIN32_LEAN_AND_MEAN)
 - fixed a bug where the tracker wouldn't expire some peers

Beta 7.2
 - more optimizations (startup should be even faster)

src/bnbt.cpp
/***
*
* BNBT Beta 8.5 - A C++ BitTorrent Tracker
* Copyright (C) 2003-2005 Trevor Hogan
*
* This library is free software; you can redistribute it and/or
* modify it under the terms of the GNU Lesser General Public
* License as published by the Free Software Foundation; either
* version 2.1 of the License, or (at your option) any later version.
*
* This library is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
* Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public
* License along with this library; if not, write to the Free Software
* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*
***/

#include <signal.h>

#include "bnbt.h"
#include "bnbt_mysql.h"
#include "config.h"
#include "server.h"
#include "tracker.h"

#ifdef WIN32
 #include "util_ntservice.h"
#endif

#include "util.h"

#ifndef WIN32
 int GetLastError( ) { return errno; }
#endif

const char *GetLastErrorString( )
{
#ifdef WIN32
	return UTIL_ErrorToString( GetLastError( ) );
#else
	return UTIL_ErrorToString( errno );
#endif
}

time_t giStartTime;

unsigned long GetTime( )
{
	return (unsigned long)( time( NULL ) - giStartTime );
}

CServer *gpServer;
CTracker *gpTracker;
string gstrErrorLogDir;
string gstrErrorLogFile;
FILE *gpErrorLog;
string gstrAccessLogDir;
string gstrAccessLogFile;
FILE *gpAccessLog;
unsigned long giErrorLogCount;
unsigned long giAccessLogCount;
int giFlushInterval;
bool gbDebug;
unsigned int giMaxConns;
unsigned int giMaxRecvSize;
string gstrStyle;
string gstrCharSet;
string gstrRealm;

void sigCatcher( int sig )
{
	signal( SIGABRT, sigCatcher );
	signal( SIGINT, sigCatcher );
	signal( SIGTERM, sigCatcher ); /* =X= */

	switch( sig )
	{
	case SIGABRT:
		UTIL_LogPrint( "BNBT caught SIGABRT\n" );

		break;
	case SIGINT:
		UTIL_LogPrint( "BNBT caught SIGINT\n" );

		break;
	case SIGTERM:
		UTIL_LogPrint( "BNBT caught SIGTERM\n" );

		break;
	default:
		UTIL_LogPrint( "BNBT caught unknown signal\n" );
	} /* =X= */

	if( gpServer )
	{
		if( gpServer->isDying( ) )
			exit( 1 );
		else
			gpServer->Kill( );
	}
	else
		exit( 1 );
}

int main( int argc, char *argv[] )
{
#ifdef WIN32
	if( argv[0] )
	{
		char *szEndPos = strrchr( argv[0], '\\' );

		if( szEndPos )
		{
			char *szEXEPath = new char[szEndPos - argv[0] + 1];
			memcpy( szEXEPath, argv[0], szEndPos - argv[0] );
			szEXEPath[szEndPos - argv[0]] = '\0';

			SetCurrentDirectory( szEXEPath );

			delete [] szEXEPath;
		}
	}

	if( argc > 1 )
	{
		if( _stricmp( argv[1], "-i" ) == 0 )
		{
			// install service

			if( UTIL_NTServiceTest( ) )
				printf( "BNBT Service is already installed!\n" );
			else
			{
				if( UTIL_NTServiceInstall( ) )
					printf( "BNBT Service installed.\n" );
				else
					printf( "BNBT Service failed to install (error %d).\n", GetLastError( ) );
			}

			return 0;
		}
		else if( _stricmp( argv[1], "-u" ) == 0 )
		{
			// uninstall service

			if( !UTIL_NTServiceTest( ) )
				printf( "BNBT Service is not installed!\n" );
			else
			{
				if( UTIL_NTServiceUninstall( ) )
					printf( "BNBT Service uninstalled.\n" );
				else
					printf( "BNBT Service failed to uninstall (error %d).\n", GetLastError( ) );
			}

			return 0;
		}
		else if( _stricmp( argv[1], "-start" ) == 0 )
		{
			// start

			if( !UTIL_NTServiceTest( ) )
				printf( "BNBT Service is not installed!\n" );
			else
			{
				printf( "Starting BNBT Service.\n" );

				if( !UTIL_NTServiceStart( ) )
					printf( "BNBT Service failed to start (error %d).\n", GetLastError( ) );
			}

			return 0;
		}
		else if( _stricmp( argv[1], "-stop" ) == 0 )
		{
			// stop

			if( !UTIL_NTServiceTest( ) )
				printf( "BNBT Service is not installed!\n" );
			else
			{
				printf( "Stopping BNBT Service.\n" );

				if( !UTIL_NTServiceStop( ) )
					printf( "BNBT Service failed to stop (error %d).\n", GetLastError( ) );
			}

			return 0;
		}
		else if( _stricmp( argv[1], "-s" ) == 0 )
		{
			// internal start

			SERVICE_TABLE_ENTRY st[] = {
				{ BNBT_SERVICE_NAME, NTServiceMain },
				{ NULL, NULL }
			};

			StartServiceCtrlDispatcher( st );

			return 0;
		}
	}
#else // Not Windows, no command line arguments.
	(void) argc; // Avoid unused argument warning.
	(void) argv;

	// disable SIGPIPE since some systems like OS X don't define MSG_NOSIGNAL

	signal( SIGPIPE, SIG_IGN );
#endif

	// catch SIGABRT and SIGINT

	signal( SIGABRT, sigCatcher );
	signal( SIGINT, sigCatcher );
	signal( SIGTERM, sigCatcher ); /* =X= */

	return bnbtmain( );
}

int bnbtmain( )
{
	srand( (unsigned int)time( NULL ) );

	giStartTime = time( NULL );

	CFG_Open( CFG_FILE );
	CFG_SetDefaults( );
	CFG_Close( CFG_FILE );

	gstrErrorLogDir = CFG_GetString( "bnbt_error_log_dir", string( ) );

	if( !gstrErrorLogDir.empty( ) && gstrErrorLogDir[gstrErrorLogDir.size( ) - 1] != PATH_SEP )
		gstrErrorLogDir += PATH_SEP;

	gpErrorLog = NULL;
	gstrAccessLogDir = CFG_GetString( "bnbt_access_log_dir", string( ) );

	if( !gstrAccessLogDir.empty( ) && gstrAccessLogDir[gstrAccessLogDir.size( ) - 1] != PATH_SEP )
		gstrAccessLogDir += PATH_SEP;

	gpAccessLog = NULL;
	giErrorLogCount = 0;
	giAccessLogCount = 0;
	giFlushInterval = CFG_GetInt( "bnbt_flush_interval", 100 );
	gbDebug = CFG_GetInt( "bnbt_debug", 1 ) == 0 ? false : true;
	giMaxConns = CFG_GetInt( "bnbt_max_conns", 64 );
	giMaxRecvSize = CFG_GetInt( "bnbt_max_recv_size", 524288 );
	gstrStyle = CFG_GetString( "bnbt_style_sheet", string( ) );
	gstrCharSet = CFG_GetString( "bnbt_charset", "iso-8859-1" );
	gstrRealm = CFG_GetString( "bnbt_realm", "BNBT" );

	// start winsock

#ifdef WIN32
	WSADATA wsadata;

	if( WSAStartup( MAKEWORD( 2, 0 ), &wsadata ) != 0 )
	{
		UTIL_LogPrint( "error - unable to start winsock (error %s)\n", GetLastErrorString( ) );

		if( gpAccessLog )
			fclose( gpAccessLog );

		if( gpErrorLog )
			fclose( gpErrorLog );

		return 1;
	}
#endif

	// start mysql

#ifdef BNBT_MYSQL
	if( !( gpMySQL = mysql_init( NULL ) ) )
	{
		UTIL_LogPrint( "mysql error - %s\n", mysql_error( gpMySQL ) );

		if( gpAccessLog )
			fclose( gpAccessLog );

		if( gpErrorLog )
			fclose( gpErrorLog );

		return 1;
	}

	gstrMySQLHost = CFG_GetString( "mysql_host", string( ) );
	gstrMySQLDatabase = CFG_GetString( "mysql_database", "bnbt" );
	gstrMySQLUser = CFG_GetString( "mysql_user", string( ) );
	gstrMySQLPassword = CFG_GetString( "mysql_password", string( ) );
	giMySQLPort = CFG_GetInt( "mysql_port", 0 );

	if( !( mysql_real_connect( gpMySQL, gstrMySQLHost.c_str( ), gstrMySQLUser.c_str( ), gstrMySQLPassword.c_str( ), gstrMySQLDatabase.c_str( ), giMySQLPort, NULL, 0 ) ) )
	{
		UTIL_LogPrint( "mysql error - %s\n", mysql_error( gpMySQL ) );

		if( gpAccessLog )
			fclose( gpAccessLog );

		if( gpErrorLog )
			fclose( gpErrorLog );

		return 1;
	}

	UTIL_LogPrint( "mysql - connected\n" );

	// CMySQLQuery mq( "REPLACE INTO bnbt_status (status_key,status_datetime_value) VALUES('bnbt_start_time',NOW())" );
#endif

	// Fix for bugs relating to clients connecting before the tracker processor is constructed - DWK in consultation with Trevor
	gpServer = NULL;
	gpTracker = new CTracker( );
	gpServer = new CServer( );

	while( 1 )
	{
		if( gpServer->Update( true ) )
		{
			delete gpServer;
			delete gpTracker;

			gpServer = NULL;
			gpTracker = NULL;

			break;
		}

		// tphogan - untied CTracker from CServer so don't forget to update on frame
		// note that CTracker updates are just long interval updates
		// stuff like parsing the allowed dir and refreshing the static files
		// this can't fail so no need to check as with CServer above

		gpTracker->Update( );
	}

#ifdef BNBT_MYSQL
	mysql_close( gpMySQL );
#endif

	if( gpAccessLog )
		fclose( gpAccessLog );

	if( gpErrorLog )
		fclose( gpErrorLog );

#ifdef WIN32
	WSACleanup( );
#endif

	return 0;
}
